src/R2DH.jl

```diff
@@ -0,0 +1,331 @@
+    * `fk`: the value of the smooth objective at the final iterate
+    * `hk`: the value of the nonsmooth objective at the final iterate
+    * `sqrt_ξ_νInv`: the square root of the ratio of the nonsmooth term to the regularization parameter
+    * `elapsed_time`: the elapsed time
+"""
+function R2DH(
+  f::F,
+  ∇f!::G,
+  h::H,
+  D::DQN,
+  options::ROSolverOptions{R},
+  x0::AbstractVector{R};
+  Mmonotone::Int = 6,
+  selected::AbstractVector{<:Integer} = 1:length(x0),
+  kwargs...,
+) where {F <: Function, G <: Function, H, R <: Real, DQN <: AbstractDiagonalQuasiNewtonOperator}
+  start_time = time()
+  elapsed_time = 0.0
+  ϵ = options.ϵa
+  ϵr = options.ϵr
+  neg_tol = options.neg_tol
+  verbose = options.verbose
+  maxIter = options.maxIter
+  maxTime = options.maxTime
+  σmin = options.σmin
+  σk = options.σk
+  η1 = options.η1
+  η2 = options.η2
+  ν = options.ν
+  γ = options.γ
+  θ = options.θ
+
+  local l_bound, u_bound
+  has_bnds = false
+  for (key, val) in kwargs
+    if key == :l_bound
+      l_bound = val
+      has_bnds = has_bnds || any(l_bound .!= R(-Inf))
+    elseif key == :u_bound
+      u_bound = val
+      has_bnds = has_bnds || any(u_bound .!= R(Inf))
+    end
+  end
+
+  if verbose == 0
+    ptf = Inf
+  elseif verbose == 1
+    ptf = round(maxIter / 10)
+  elseif verbose == 2
+    ptf = round(maxIter / 100)
+  else
+    ptf = 1
+  end
+
+  # initialize parameters
+  xk = copy(x0)
+  hk = h(xk[selected])
+  if hk == Inf
+    verbose > 0 && @info "R2DH: finding initial guess where nonsmooth term is finite"
+    prox!(xk, h, x0, one(eltype(x0)))
+    hk = h(xk[selected])
+    hk < Inf || error("prox computation must be erroneous")
+    verbose > 0 && @debug "R2DH: found point where h has value" hk
+  end
+  hk == -Inf && error("nonsmooth term is not proper")
+
+  xkn = similar(xk)
+  s = zero(xk)
+  ψ = has_bnds ? shifted(h, xk, l_bound - xk, u_bound - xk, selected) : shifted(h, xk)
+
+  Fobj_hist = zeros(maxIter+1)
+  Hobj_hist = zeros(maxIter+1)
+  time_hist = zeros(maxIter+1)
+  FHobj_hist = fill!(Vector{R}(undef, Mmonotone - 1), R(-Inf))
+  Complex_hist = zeros(Int, maxIter+1)
+  k_prox = 0
+  if verbose > 0
+    #! format: off
+    @info @sprintf "%6s %8s %8s %7s %8s %7s %7s %7s %1s" "iter" "f(x)" "h(x)" "√(ξ/ν)" "ρ" "σ" "‖x‖" "‖s‖" ""
+    #! format: off
+  end
+
+  local ξ
+  k = 0
+
+  fk = f(xk)
+  ∇fk = similar(xk)
+  ∇f!(∇fk, xk)
+  ∇fk⁻ = copy(∇fk) 
+  spectral_test = isa(D, SpectralGradient)
+  Dkσk = D.d .+ σk
+  DNorm = norm(D.d, Inf)
+
+  ν = 1 / ((DNorm + σk) * (1 + θ))
+  mν∇fk = -ν * ∇fk
+  sqrt_ξ_νInv = one(R)  
+
+  optimal = false
+  tired = maxIter > 0 && k ≥ maxIter || elapsed_time > maxTime
+
+  while !(optimal || tired)
+    k_prox += 1
+    # model with diagonal hessian
+    φ(d) = ∇fk' * d + (d' * (Dkσk .* d)) / 2
+    mk(d) = φ(d) + ψ(d)
+
+    if spectral_test
+      prox!(s, ψ, mν∇fk, ν)
```

Comment:
This doesn’t correspond to R2DH; I think this should use `ν = 1 / (DNorm + σk)`.

Reply:
We have to prove that this version of R2DH converges, but it should be similar to TRDH.

Reply:
I thought about it a while ago, but in our analysis, which assumes that \(D_k\) might be unbounded, it becomes more difficult, and TRDH relies on the boundedness of \(D_k\) if I'm not mistaken.